test/model/typing_status_test.dart

```diff
@@ -0,0 +1,278 @@
+import 'dart:async';
+
+import 'package:checks/checks.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:zulip/api/model/events.dart';
+import 'package:zulip/model/narrow.dart';
+import 'package:zulip/model/typing_status.dart';
+
+import '../example_data.dart' as eg;
+import '../fake_async.dart';
+
+extension TypingStatusChecks on Subject<TypingStatus> {
+  Subject<Map<SendableNarrow, Map<int, Timer>>> get typistsByLocation => has((x) => x.typistIdsByNarrow, 'typistsByLocation');
+}
+
+extension TimerChecks on Subject<Timer> {
+  Subject<bool> get isActive => has((x) => x.isActive, 'isActive');
+}
+
+
+void main() {
+  late TypingStatus model;
+  late int notifiedCount;
+
+  void prepareModel() {
+    model = TypingStatus(
+        selfUserId: eg.selfUser.userId,
+        typingStartedExpiryPeriod: const Duration(milliseconds: 15000),
+        typingStoppedWaitPeriod: const Duration(milliseconds: 5000),
+        typingStartedWaitPeriod: const Duration(milliseconds: 10000));
+    check(model).typistsByLocation.isEmpty();
+    notifiedCount = 0;
+    model.addListener(() => notifiedCount += 1);
+  }
+
+  void checkNotNotified() {
+    check(notifiedCount).equals(0);
+  }
+
+  void checkNotifiedOnce() {
+    check(notifiedCount).equals(1);
+    notifiedCount = 0;
+  }
+
+  TypingEvent mkTypingEvent(SendableNarrow narrow, TypingOp op, int senderId) {
+    switch (narrow) {
+      case TopicNarrow():
+        return TypingEvent(id: 1, op: op, senderId: senderId,
+          messageType: MessageType.stream,
+          streamId: narrow.streamId,
+          topic: narrow.topic,
+          recipientIds: null);
+      case DmNarrow():
+        return TypingEvent(id: 1, op: op, senderId: senderId,
+          messageType: MessageType.private,
+          recipientIds: narrow.allRecipientIds,
+          streamId: null,
+          topic: null);
+    }
+  }
+
+  final stream = eg.stream();
+  final topicNarrow = TopicNarrow(stream.streamId, 'foo');
+
+  final otherUser = eg.otherUser;
+  final thirdUser = eg.thirdUser;
+  final dmNarrow = DmNarrow.withUser(otherUser.userId, selfUserId: eg.selfUser.userId);
+  final groupNarrow = DmNarrow.withOtherUsers(
+    [otherUser.userId, thirdUser.userId], selfUserId: eg.selfUser.userId);
+
+  anyTimer(Subject<Object?> it) => it.isA<Timer>();
+
+  group('handle typing start events', () {
+    test('add typists in separate narrows', () {
+      prepareModel();
+
+      model.handleTypingEvent(
+        mkTypingEvent(dmNarrow, TypingOp.start, otherUser.userId));
```

Comment:
There's a lot of repetition of

```dart
      model.handleTypingEvent(
        mkTypingEvent(
```

, which always takes at least two lines. Perhaps worth making a helper that lets us do the same thing in one line?